From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?I=C3=B1aki=20Amatria=20Barral?= <inaki.amatria@appentra.com>
Date: Wed, 2 Apr 2025 09:20:20 +0200
Subject: Stop parsing continuation lines in number literals

The previous algorithm incorrectly handled line continuations in number
literals. Instead of attempting to parse them incorrectly, we now treat
them as a parsing error. This allows library users to handle such cases
explicitly.
---
 src/scanner.c                      | 104 +++++++++++++++++++----------
 test/corpus/expressions.txt        |  31 ++++-----
 test/corpus/line_continuations.txt |   3 +-
 3 files changed, 85 insertions(+), 53 deletions(-)

diff --git a/src/scanner.c b/src/scanner.c
index 846a5d8..4b1a626 100644
--- a/src/scanner.c
+++ b/src/scanner.c
@@ -41,6 +41,7 @@ typedef struct {
 // for example: we need to check label stack first before we can decide whether
 // we have a do-label or an integer literal
 typedef enum {
+    NUMBER_ERROR,
     NUMBER_NONE,
     NUMBER_INTEGER,
     NUMBER_FLOAT
@@ -56,6 +57,16 @@ typedef struct {
     int digit_count;
 } NumberResult;
 
+typedef enum {
+    False,
+    True,
+    Error,
+} BoolOrErr;
+
+static BoolOrErr bool_or_err_max(BoolOrErr lhs, BoolOrErr rhs) {
+    return lhs >= rhs ? lhs : rhs;
+}
+
 //  consume current character into current token and advance
 static inline void advance(TSLexer *lexer) { lexer->advance(lexer, false); }
 
@@ -112,9 +123,9 @@ static bool skip_literal_continuation_sequence(TSLexer *lexer) {
 }
 
 // consume digits and compute value if requested
-static bool scan_int(TSLexer *lexer, int32_t *value, int *count) {
+static BoolOrErr scan_int(TSLexer *lexer, int32_t *value, int *count) {
     if (!iswdigit(lexer->lookahead)) {
-        return false;
+        return False;
     }
 
     if (value && count) {
@@ -122,31 +133,32 @@ static bool scan_int(TSLexer *lexer, int32_t *value, int *count) {
         *count = 0;
     }
 
-    // outer loop for handling line continuations
-    while (true) {
-        // consume digits
-        while (iswdigit(lexer->lookahead)) {
-            if (value && count && *count < 7) {
-                *value = *value * 10 + (lexer->lookahead - '0');
-                (*count)++;
-            }
-            advance(lexer);
+    // consume digits
+    while (iswdigit(lexer->lookahead)) {
+        if (value && count && *count < 7) {
+            *value = *value * 10 + (lexer->lookahead - '0');
+            (*count)++;
         }
-        lexer->mark_end(lexer);
+        advance(lexer);
+    }
+    lexer->mark_end(lexer);
 
-        if (lexer->lookahead == '&') {
-            // with the lookahead check above, it returns true if in a continuation
-            // line and consumes both & and whitespace characters inbetween
-            if (skip_literal_continuation_sequence(lexer)) {
-                continue;
-            }
-        }
-
-        // no digits and no continuation found
-        break;
+    // Return an error if a line continuation is found. This scanner cannot
+    // handle line continuations, particularly in cases like:
+    //
+    // ```fortran
+    // b = 6& ! foo
+    //   &7;
+    // ```
+    //
+    // Here, the scanner would need to return multiple tokens, but tree-sitter
+    // expects only a single token.
+    if (lexer->lookahead == '&') {
+        return Error;
     }
 
-    return true;
+    // no more digits and no continuation found
+    return True;
 }
 
 /// Scan integer or float of the forms 1XXX, 1.0XXX, 0.1XXX, 1.XDX, .1X etc.
@@ -158,14 +170,14 @@ static NumberResult scan_number(TSLexer *lexer) {
 
     // collect initial digits and compute value (specifically required to
     // determine label value)
-    bool digits = scan_int(lexer, &result.value, &result.digit_count);
+    BoolOrErr digits = scan_int(lexer, &result.value, &result.digit_count);
 
     if (lexer->lookahead == '.') {
         advance(lexer);
         // exclude decimal if followed by any letter other than d/D and e/E
         // if no leading digits are present and a non-digit follows
         // the decimal it's a nonmatch.
-        if (digits && !iswalnum(lexer->lookahead)) {
+        if ((digits == True) && !iswalnum(lexer->lookahead)) {
             lexer->mark_end(lexer); // add decimal to token
         }
         // this is not yet decided, we still need to find some digits
@@ -175,9 +187,9 @@ static NumberResult scan_number(TSLexer *lexer) {
     // if next char isn't number return since we handle exp
     // notation and precision identifiers separately. If there are
     // no leading digit it's a nonmatch.
-    digits = scan_int(lexer, NULL, NULL) || digits;
+    digits = bool_or_err_max(scan_int(lexer, NULL, NULL), digits);
 
-    if (digits) {
+    if (digits == True) {
         // process exp notation
         if (is_exp_sentinel(lexer->lookahead)) {
             advance(lexer);
@@ -185,17 +197,26 @@ static NumberResult scan_number(TSLexer *lexer) {
                 advance(lexer);
                 lexer->mark_end(lexer);
             }
-            if (!scan_int(lexer, NULL, NULL)) {
+            switch (scan_int(lexer, NULL, NULL)) {
+            case Error:
+                result.type = NUMBER_ERROR;
+                return result; // invalid number
+            case True:
                 result.type = NUMBER_INTEGER;
                 return result; // valid number token with junk after it
+            case False:
+                break;
             }
             result.type = NUMBER_FLOAT;
         }
     }
 
-    if (!digits) {
+    if (digits == False) {
         result.type = NUMBER_NONE;
     }
+    if (digits == Error) {
+        result.type = NUMBER_ERROR;
+    }
     return result;
 }
 
@@ -542,37 +563,41 @@ static bool scan_do_label_continue(Scanner *scanner, TSLexer *lexer, int32_t lab
 }
 
 // check for label, number of boz token
-static bool scan_label_number_boz(Scanner *scanner, TSLexer *lexer, const bool *valid_symbols) {
+static BoolOrErr scan_label_number_boz(Scanner *scanner, TSLexer *lexer,
+                                       const bool *valid_symbols) {
     // extract out root number from expression (without kind if present)
     NumberResult result = scan_number(lexer);
+    if (result.type == NUMBER_ERROR) {
+        return Error;
+    }
 
     // check for a do-label, should have at most 5 digits and DO_LABEL
     // or DO_LABEL_CONTINUE are valid symbols
     if (result.type == NUMBER_INTEGER && result.digit_count < 6) {
         if (valid_symbols[DO_LABEL]) {
             scan_do_label(scanner, lexer, result.value);
-            return true;
+            return True;
         }
         if (valid_symbols[DO_LABEL_CONTINUE] &&
             scan_do_label_continue(scanner, lexer, result.value)) {
-            return true;
+            return True;
         }
     }
 
     // not a label
     if (result.type == NUMBER_INTEGER) {
         lexer->result_symbol = INTEGER_LITERAL;
-        return true;
+        return True;
     } else if (result.type == NUMBER_FLOAT) {
         lexer->result_symbol = FLOAT_LITERAL;
-        return true;
+        return True;
     }
 
     if (scan_boz(lexer)) {
-        return true;
+        return True;
     }
 
-    return false;
+    return False;
 }
 
 static bool scan(Scanner *scanner, TSLexer *lexer, const bool *valid_symbols) {
@@ -629,8 +654,13 @@ static bool scan(Scanner *scanner, TSLexer *lexer, const bool *valid_symbols) {
         valid_symbols[BOZ_LITERAL] ||
         valid_symbols[DO_LABEL] ||
         valid_symbols[DO_LABEL_CONTINUE]) {
-        if (scan_label_number_boz(scanner, lexer, valid_symbols)) {
+        switch (scan_label_number_boz(scanner, lexer, valid_symbols)) {
+        case False:
+            break;
+        case True:
             return true;
+        case Error:
+            return false;
         }
     }
 
diff --git a/test/corpus/expressions.txt b/test/corpus/expressions.txt
index c6b7225..623668b 100644
--- a/test/corpus/expressions.txt
+++ b/test/corpus/expressions.txt
@@ -1048,9 +1048,11 @@ end program
     (variable_declaration
       (intrinsic_type)
       (type_qualifier)
-      (init_declarator
-        (identifier)
-        (number_literal)))
+      (identifier))
+        (ERROR
+          (ERROR
+            (number_literal))
+          (statement_label_reference))
     (end_of_statement)
     (variable_declaration
       (intrinsic_type)
@@ -1060,18 +1062,17 @@ end program
       (init_declarator
         (identifier)
         (array_literal
-          (number_literal)
-          (number_literal))))
-    (end_of_statement)
-    (variable_declaration
-      (intrinsic_type)
-      (type_qualifier
-        (argument_list
-          (number_literal)))
-      (init_declarator
-        (identifier)
-        (array_literal
-          (number_literal)
+          (ERROR
+            (number_literal))
+          (identifier)
+          (call_expression
+            (identifier)
+            (argument_list
+              (number_literal)))
+          (ERROR
+            (ERROR
+              (number_literal))
+            (number_literal))
           (number_literal))))
     (end_of_statement)
     (end_program_statement)
diff --git a/test/corpus/line_continuations.txt b/test/corpus/line_continuations.txt
index a5ae1d1..46c2e04 100644
--- a/test/corpus/line_continuations.txt
+++ b/test/corpus/line_continuations.txt
@@ -339,7 +339,8 @@ end program test
         (array_literal
           (number_literal)
           (number_literal)
-          (number_literal))))
+          (ERROR
+            (number_literal)))))
     (end_of_statement)
     (end_program_statement
       (name))
